File: namaste_dsa/BinarySearch/BinarySearchOnAnswers/KthMinimumElementInSortedMatrix.js
function findKthMinimum(matrix, k) {
    let n = matrix.length

    function countSmaller(x) {
        let row = n - 1
        let col = 0
        let count = 0

        while (row >= 0 && col < n) {
            if (matrix[row][col] <= x) {
                count += row + 1
                col++
            }
            else {
                row--
            }
        }

        return count
    }

    let low = matrix[0][0]
    let high = matrix[n - 1][n - 1]

    while (low < high) {
        let mid = low + Math.floor((high - low) / 2)

        if (countSmaller(mid) < k) {
            low = mid + 1
        }
        else {
            high = mid
        }
    }

    return low
}

File: namaste_dsa/BinarySearch/BinarySearchOnAnswers/KthMinimumElementInSortedMatrix.py
class Solution(object):
    def kthSmallest(self, matrix, k):
        n = len(matrix)

        def countLessEqual(x):
            count = 0
            row = n - 1
            col = 0

            while row >= 0 and col < n:
                if matrix[row][col] <= x:
                    count += row + 1
                    col += 1
                else:
                    row -= 1
            return count

        low = matrix[0][0]
        high = matrix[n -1][n - 1]

        while low < high:
            mid = low + (high - low) // 2

            if countLessEqual(mid) < k:
                low = mid + 1
            else:
                high = mid
        
        return low